DataAccessPlugin Completion Report Enhancement

capricornunicorn123 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
![Heading Image](https://github.com/constellation-app/constellation/assets/139201011/9bbc4f76-dac5-4ef2-a723-d3c82fec7e65)

This pull request addresses issues with `DataAccessPlugins` and their ability to provide contextually meaningful process completion information to the `PluginReporterView`. All plugins within Core that implement the `DataAccessPlugin` were included. 

**Change 1**
All `PluginReports` in the `PluginReporter` are now expandable, showing all past progress updates made during the plugin execution. 

**Change 2**
Plugin report information has been broken into two flavors from a software development perspective:
- *Execution information* is updated with the `setExecutionStage(...)` method and updates the plugin reporter of information on plugin execution stages throughout its execution life cycle. These stages can be updated from any menthod that has access to a `PluginInteraction` but should be called on the higher level such as within `run()` methods. Updating at the time that methods like `read()`, `query()` and `edit()` are called would be appropriate. This information is useful during long running plugins but is intended to only persist during plugin execution and be hidden after the plugin has completed successfully.

- *Progress information* is set using the `setProgress(...)` method and is used to update information specific to data manipulation. This information appears as it is created and is displayed along with the execution information. This information is intended to persist after the completion of the plugin and should inform the user of the actions performed by the plugin.

**Change 3**
To further aid in the understanding of a plugins progress, a `ProgressBar` was enabled. Progress bars have two modes of operation:

- *Indeterminant* progress bars should be used by default when a process is running. The progress bar moves continually until the process is complete. 
![Indeterminent](https://github.com/constellation-app/constellation/assets/139201011/959a788e-37c7-488b-95cb-718340a3190a)


- *Determinant* progress bars should be used within `read()`, `query()` and `edit()` methods where possible. The progress bar moves discretely representing the current step of the process with regards to the total anticipated steps required. 
![Determinent](https://github.com/constellation-app/constellation/assets/139201011/053c06e9-b229-44f4-835e-f0f34d21791c)


**Change 4**
To better inform users of the state of the plugin (if it is Waiting, Running or Complete) the plugin running state has been expressed in the title of the `PluginReport`
![states](https://github.com/constellation-app/constellation/assets/139201011/89e232f9-280a-41e4-bd49-a127d607383e)


**Detailed Change Infromation**
During exploration it was noted that `DataAccessPlugins` report to the `PluginReporterView`, facilitated with a `PluginInteraction` via the following extensions of `AbstractPlugin`:
- `SimplePlugin`
- `SimpleEditPlugin`
- `SimpleQueryPlugin`
- `WorkflowQueryPlugin`
- `RecordStorePlugin`

Each of these plugin flavors provide different processing pathways which can include methods such as `edit( )`, `query( )`, `execute( )` and `run( )`. Ultimately, a number of these plugins provided default reports to the `PluginInteraction` after the implementation of the `DataAccessPlugin` extending upon it. This resulted in an override of detailed reports implemented by these `DataAccessPlugins`. The following steps were taken to rectify this issue:

- Refactor the `PluginReporterView` to enable plugin reports to express all reports called over their lifetime.
- Refractor `PluginReport` to store Execution Information and Progress Information separately.
- Refactor abstract plugins to report their progress using `setExecutionStage()`
- Ensure `DataAccessPlugins` reports are relevant in terms of content and quantity to the `PluginReporterView`.

The completion report for the following views that extend the `SimpleQueryPlugin` have been considered:
- `MergeNodesPlugin`
- `MergeTransactionsPlugin`
- `RemoveNodesPlugin`
- `ExtractWordsFromTextPlugin`
- `SelectAllPlugin`
- `SelectTopNPlugin`

The completion report for the following views that extend the `RecordStoreQueryPlugin` have been considered:
- `TestParametersPlugin`
- `ExtractTypesFromTextPlugin`
- `ImportGraphFilePlugin`


The completion report for the following views that extend the `SimpleEditPlugin` have been considered:
- `RemoveUnusedAttributesPlugin`
- `SplitNodesPlugin`
- `CompleteSchemaPlugin`

The completion report for the following view that extends the `WorkflowQueryPlugin` has been considered:
- `TestParametersBatched`

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs

The lack of hierarchy in abstract plugin types means that default reports made by abstract classes can include irrelevant information. i.e. a plugin that only reads data yet extends the simple query plugin will have automated reports such as "Querying..." and "Editing...". Although this ticket only deals with the reports made by a select number of plugins, the topic of tightening the abstract plugin hierarchy was discussed to standardize the plugin reports by ensuring the inherited classes are specific to their function. 

<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
This issue effects a core module.
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits

<!-- What benefits will be realized by the code change? -->
Users will now be able to view contextual information form the `PluginReporter`. 

Whilst a plugin is running users can understand what stage the plugin is at and understand if the plugin is executing or waiting and see how ling is left for this process to complete.

Once the process is finished users can understand what specifically the process has completed and will not be able to see irrelevant information relating to the execution of the plugin.
 
### Possible Drawbacks

**Noted Potential Anomalies**

*SelectTopNPlugin*
Essentially, N Nodes are selected from the nodes that directly neighbour a selection based on their level of connectedness. In cases where more than one node is selected, the neighboring nodes being selected in this plugin may include nodes already in the current selection. The reporting functionality of this plugin will report the number of nodes that the plugin has identified as needing to be selected regardless of whether these currently selected or not. Use Case - The user has two nodes selected out of 100 nodes and wants to select the 3 most connected neighboring nodes using the SelectTopNPlugin. Currently after running the plugin the user may have between 3 and 5 total nodes selected (the two current nodes + between 1 and 3 additional nodes depending on whether the currently selected nodes were the most connected nodes to each other compared to all other neighboring nodes). The Select Top N function will however, have successfully identified three nodes and not factoring in whether they were a part of the original selection and will report that 3 nodes were selected in all cases. 

Is this SelectTopN functionality correct? Should the Top N be from non-selected nodes? Should the report only report on newly selected nodes? Should the report, report on all selected nodes?

*Report Quantity*
Due to the tight coupling of report content and DataAccessView progress control a number of report messages will be displayed in some detailed reports of DataAccessViews that may seem relevant. i.e. "Reading...", "waiting for other plugins...". Is this information still relevant to plugin reports?

*SplitNodesPlugin*
The SplitNodesPlugin takes a selection of nodes and an input character (or set of characters) and attempts to split the nodes on occurrences of that character in their IDs. The newly created Nodes will have Id's of the text either side of the occurrences of the split character(s). i.e "node_3" split on character "_" will create node "node" and node"3". If a newly created node has a conflicting ID it will inherit the transactions of the conflicting node and override the conflicting nodes attributes. This doesn't seem right... If this is right, this is also difficult to log accurately. 

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#1858
<!-- Link any applicable issues here -->

OrionsGuardian commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveUnusedAttributesPlugin.java:
```diff
@@ -68,15 +68,23 @@ public int getPosition() {
+            
+            //Setup progress report controll data
+            int currentProgress = 0;
+            int maxProgress = graph.getAttributeCount(GraphElementType.VERTEX) + graph.getAttributeCount(GraphElementType.TRANSACTION);
```
`final`

OrionsGuardian commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/importing/ExtractWordsFromTextPlugin.java:
```diff
@@ -373,7 +377,7 @@ public void edit(final GraphWriteMethods wg, final PluginInteraction interaction
                         final ZonedDateTime datetime = wg.getObjectValue(transactionDatetimeAttributeId, transactionId);
 
-                        matched.forEach(word -> {
+                        for (String word : matched ){
```
`final` and remove extra space

antares1470 replied:
can you add final here as well?

OrionsGuardian commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/importing/ImportGraphFilePlugin.java:
```diff
@@ -153,7 +153,7 @@ protected RecordStore query(final RecordStore query, final PluginInteraction int
         }
         
-        interaction.setProgress(1, 0, "Completed successfully - added " + result.size() + " entities.", true);
+        interaction.setProgress(1, 0, "Imported " + result.size() + " entity(ies).", true);
```
Not sure I like the look of this: "entity(ies)" 
Can you put in a boolean check to show either "entity" or "entities" as appropriate.

OrionsGuardian commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -47,7 +47,7 @@ public class PluginReport {
     private long stopTime = -1;
 
-    private String message = null;
+    private final List<String> messageLog = new ArrayList<>();
```
When changing the member variable definition, you should also be changing the getter/setting methods to apply to the new type, and appropriately named functions when accessing the type.
ie. getMessageLog(), setMessageLog(), addMessage(), getLastMessage(), etc.
Will likely need some refactoring of calls to this class.

capricornunicorn123 replied:
Have refactored code in line with comments. Note, have not implemented function setMessageLog() as the intended use of the attribute is to be an unchangeable record that is created at the start of the PluginReport and to persist for it's lifetime. the setMessageLog() would provide the possibility to override the mesageLog and contradict its purpose.

OrionsGuardian commented on CoreVisualSchema/src/au/gov/asd/tac/constellation/graph/schema/visual/plugins/CompleteSchemaPlugin.java:
```diff
@@ -40,25 +40,32 @@ public class CompleteSchemaPlugin extends SimpleEditPlugin {
             int currentProgress = 0;
-            int maxProgress = graph.getVertexCount() + graph.getTransactionCount();
-
+            int vertexCount = graph.getVertexCount();
+            int transactionCount = graph.getTransactionCount();
+            int maxProgress = vertexCount + transactionCount;
```
`final` x 3

antares1470 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/importing/ExtractWordsFromTextPlugin.java:
```diff
@@ -281,6 +281,7 @@ public void edit(final GraphWriteMethods wg, final PluginInteraction interaction
             throw new PluginException(PluginNotificationLevel.ERROR, msg);
         }
 
+        Set<String> newNodes = new HashSet<>();
```
can you make this final?

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -347,4 +348,23 @@ public PluginReport addChildReport(Plugin plugin) {
+     * this report.
+     */
+    public String getMessageLog() {
+        String messageLogString = "";
```
can you use a `StringBuilder` here

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/text/TextPluginInteraction.java:
```diff
@@ -60,6 +60,11 @@ public void setProgress(int currentStep, int totalSteps, String message, boolean
+    
+    @Override
+    public void setProgress(int currentStep, int totalSteps, boolean cancellable) throws InterruptedException {
+        //LOGGER.log(Level.INFO, "currentStep={0} totalSteps={1} message={2}", new Object[]{currentStep, totalSteps, message});
```
are you able to remove this commented code, or alternatively uncomment and use (will need to remove message part first though)

antares1470 commented:
@capricornunicorn123 to address some of your anomolies, I think it's best to just treat each plugin in isolation. I think trying to grab the context of the graph may make things unnecessarily complicated. e.g. with split nodes, if the split generates a node which already exists on the graph, I don't think you want to try account for the merge that may subsequently follow it (which FYI merging with the existing node is the correct behaviour as each set node has a unique set of key attributes so any nodes with the same key attributes must be the same node) as that behaviour exists outside of the plugin itself. In that case, if the node split into 3, you'd simply want to say that it split into 3 nodes and I think that is ok.

As for what is being displayed in the plugin reporter, I think you'd probably to only record the parts specific to the plugin so ones like "Editing..." and "Querying.." I don't think are particularly useful to users and might makes things confusing (e.g. Merge nodes has it presumably because it executes the query function but for a user it may seem confusing as to why and what Constellation is querying for it even though it obviously doesn't actually query anything)

capricornunicorn123 commented:
@antares1470, @OrionsGuardian 
The above comments regarding :
-	boolean checks for reporting output text plurality; and 
-	the want to remove the irrelevant "boiler plate" report messages,

relate to a situation with the plugin framework that I have been trying to understand whilst working through this ticket. I just wanted to put some points forward here as food-for-thought when considering solutions to the aforementioned issues.

Currently there is not a strong hierarchy within the framework of abstract classes plugins use to drive their functionality (the “abstract plugin” class and subsequent abstract sub classes). Most "flavors" of abstract plugin inherit directly from abstract plugin and implement their own core functionality despite sharing methods and responsibilities. This results in duplicated code throughout the abstract plugin framework. This is relevant to this ticket as these “boiler plate” report messages are currently duplicated throughout these abstract classes.

As a result, there is potential for some plugins to be implemented extending an inappropriate abstract plugin such as is the case with the ExtractWordsFromTextPlugin that extends SimpleQueryPlugin but really only needs to extend SimpleEditPlugin. 

A suggested restructure to the hierarchy to help encourage appropriate abstract plugin could look like this:

![AbstractPluginHeirachy](https://github.com/constellation-app/constellation/assets/139201011/d32d5621-88ec-4213-a497-c98ba9208d8d)
(There are other abstract plugins that aren't included in this diagram)

For now, I’ll implement your above suggestions. Let me know if you’d like to discuss the topic further and for me to give some finer detail how this restructure could look.

antares1470 commented:
@capricornunicorn123 It's worth noting that each plugin holds a slightly different purpose and slightly different needs. A `SimpleReadPlugin` for instance requires a read lock on the graph, whereas a `SimplePlugin` doesn't require a lock at all. This of course wouldn't mean they couldn't be formed into a hierarchy but its worth noting if you compare implementations of each of the `run` functions, they are slightly different, and different enough that I don't think we it would make sense to have one extend the other.

The other issue is around which functions they need implemented. `SimplePlugin` requires `execute` to be implemented whereas that function doesn't exists in others like `SimpleReadPlugin` (which requires `read` to be implemented). By making `SimpleReadPlugin` extend `SimplePlugin` instead, you would be enforcing the need for `execute` to be implemented (which is not required for any plugin bar a `SimplePlugin` and so would likely for a lot of plugins just throw an `UnsupportedOperationException`). 

The naming conventions may confuse slightly, possibly suggesting that `SimpleReadPlugin` is a subclass of `SimplePlugin`, but in the same way you probably wouldn't have chocolate ice cream being an extension of vanilla, you probably wouldn't have `SimpleReadPlugin` being an extension of `SimplePlugin` (that's how I picture it anyway, happy to hear an alternative perspective)

antares1470 commented:
Having had a longer thought on it, I do wonder whether there is much benefit to the user in seeing every step of progress in the plugin reporter. I think ultimately, the user may only be interested in the outputs of a plugin execution (which is the crux of the ticket), not necessarily caring how it got there (assuming nothing unbeknownst to the user occurred in the background during execution). The extra steps are probably most useful for developers if an issue is identified but its worth noting that if an error occurs, the plugin report will already reflect that in its output. Happy to hear other thoughts though

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveNodesPlugin.java:
```diff
@@ -96,38 +101,60 @@ public PluginParameters createParameters() {
+        final int selectedAttribute = wg.getAttribute(GraphElementType.VERTEX, VisualConcept.VertexAttribute.SELECTED.getName());
+        final int identifierAttribute = wg.getAttribute(GraphElementType.VERTEX, VisualConcept.VertexAttribute.IDENTIFIER.getName());
+        
+        // Throw an error if the relevant attributs Id's could not be found.
```
Change to "attribute IDs"

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/SplitNodesPlugin.java:
```diff
@@ -161,6 +165,141 @@ public void updateParameters(final Graph graph, final PluginParameters parameter
+        
+        totalProcessSteps = selectedVerticies.size();
+        
+        // Loop through the selected vertex's and determine how many new verticies need to be added to the graph
```
change to "vertices"

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/experimental/TestParametersPlugin.java:
```diff
@@ -368,22 +368,46 @@ protected void updateItem(final ParameterValue item, final boolean empty) {
+        
+        // No Errors thrown as parameters values are checked before being used
+        
+        // Local process-tracking varables (Process is indeteminate due to the nature of plugin reporting htrough the logger)
```
change to "variables"
change to "through"

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/utility/SelectAllPlugin.java:
```diff
@@ -25,8 +25,10 @@
+import au.gov.asd.tac.constellation.plugins.reporting.PluginReportUtilities;
 import au.gov.asd.tac.constellation.plugins.templates.PluginTags;
 import au.gov.asd.tac.constellation.plugins.templates.SimpleQueryPlugin;
+import au.gov.asd.tac.constellation.utilities.text.StringUtilities;
```
Chack that this import is not used and remove it

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/utility/SelectTopNPlugin.java:
```diff
@@ -205,81 +206,91 @@ public PluginParameters createParameters() {
+                int txId = graph.getVertexTransaction(vxId, position);
+                int sourceVxId = graph.getTransactionSourceVertex(txId);
+                int destinationVxId = graph.getTransactionDestinationVertex(txId);
+                int targetVxId = vxId == sourceVxId ? destinationVxId : sourceVxId;
```
final x4

capricornunicorn123 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/utility/SelectTopNPlugin.java:
```diff
@@ -205,81 +206,91 @@ public PluginParameters createParameters() {
+                int destinationVxId = graph.getTransactionDestinationVertex(txId);
+                int targetVxId = vxId == sourceVxId ? destinationVxId : sourceVxId;
 
+                //Tally the number of transactions between the current Node and nodes sharing a transaction
```
decapitalize "Node"

capricornunicorn123 commented on CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveNodesPluginNGTest.java:
```diff
@@ -89,13 +90,33 @@ public void testCreateParameters() {
+        
+        vertexSelectedAttribute = VisualConcept.VertexAttribute.SELECTED.ensure(graph);
+        
+        //Method should throw exception as Identifier attribute is not enabled on vertex
```
Change to "identifier"

Delphinus8821 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReportUtilities.java:
```diff
@@ -0,0 +1,81 @@
+     * @param count
+     */
+    public static final String getNodeCountString(final int count){
+        return PluginReportUtilities.toString(count, "node", "Nodes");
```
Lowercase nodes

Delphinus8821 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/utility/SelectTopNPlugin.java:
```diff
@@ -205,81 +206,91 @@ public PluginParameters createParameters() {
                 switch (mode) {
                     case NODE:
-                        destinationVertexType = graph.getObjectValue(vertexTypeAttribute, targetVxId);
+                        SchemaVertexType destinationVertexType = graph.getObjectValue(vertexTypeAttribute, targetVxId);
```
final

Delphinus8821 commented on CoreDataAccessView/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/utility/SelectTopNPlugin.java:
```diff
@@ -289,7 +300,7 @@ protected void edit(final GraphWriteMethods graph, final PluginInteraction inter
                         break;
                     case TRANSACTION:
-                        transactionType = graph.getObjectValue(transactionTypeAttribute, txId);
+                        SchemaTransactionType transactionType = graph.getObjectValue(transactionTypeAttribute, txId);
```
final

Delphinus8821 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -240,7 +250,16 @@ public Throwable getError() {
      * @param error the new error.
      */
     public void setError(Throwable error) {
```
final

Delphinus8821 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -347,4 +366,90 @@ public PluginReport addChildReport(Plugin plugin) {
+     *
+     * @return a string representing a list of all the messages in the log.
+     */
+    public String logToString(Collection<String> log){
```
final

antares1470 replied:
parameter can be final

Delphinus8821 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/text/TextPluginInteraction.java:
```diff
@@ -60,6 +61,16 @@ public void setProgress(int currentStep, int totalSteps, String message, boolean
     }
+    
+    @Override
+    public void setProgress(int currentStep, int totalSteps, boolean cancellable) throws InterruptedException {
```
finals

antares1470 replied:
parameters can be final

Delphinus8821 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/text/TextPluginInteraction.java:
```diff
@@ -60,6 +61,16 @@ public void setProgress(int currentStep, int totalSteps, String message, boolean
+    }
 
     @Override
     public void notify(PluginNotificationLevel level, String message) {
```
finals

Delphinus8821 commented on CorePluginReporterView/src/au/gov/asd/tac/constellation/views/pluginreporter/panes/PluginReportPane.java:
```diff
@@ -201,7 +208,7 @@ private void saveToClipboard() {
         try (PrintWriter out = new PrintWriter(writer)) {
```
final

OrionsGuardian commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -165,17 +170,22 @@ public String getPluginDescription() {
-        return message;
+    public String getLastMessage() {
+        if ((executionStage.equals(PluginExecutionStageConstants.COMPLETE) || executionStage.equals(PluginExecutionStageConstants.STOPPED)) && !this.messageLog.isEmpty()){
+            return this.messageLog.isEmpty() ? "" : this.messageLog.get(this.messageLog.size()-1);
```
Do not need to check for `this.messageLog.isEmpty()` here since the if condition already checked it.

antares1470 replied:
the boolean check here isn't required as the if condition preceding it means it can only enter here if `!this.messageLog.isEmpty()` is true (in other words, the boolean check here will always return false)

OrionsGuardian commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -165,17 +170,22 @@ public String getPluginDescription() {
      */
-    public void setMessage(String message) {
-        this.message = message;
+    public void addMessage(String message) {
```
`final`

antares1470 replied:
parameter can be final

antares1470 commented on CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/components/TabContextMenuNGTest.java:
```diff
@@ -60,19 +60,23 @@ public class TabContextMenuNGTest {
-            FxToolkit.registerPrimaryStage();
-        }
+     
+//        if (!FxToolkit.isFXApplicationThreadRunning()) {
```
I think this commented out code is needed as the toolkit needs to be enabled in order to work (it currently fails locally)

antares1470 commented on CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveNodesPluginNGTest.java:
```diff
@@ -89,13 +90,33 @@ public void testCreateParameters() {
     @Test
     public void testEditAttributesNotFound() throws InterruptedException, PluginException {
```
I think with the way this test has been modified, it would be better to split into 2 separate tests, one for selected missing and one for identifier missing. This would then allow you to make use of the expectedException parameter for the Test annotation instead of throwing fail if the exception doesn't occur. The 3rd case should be covered by the other tests (which in theory should run intended behaviour)

antares1470 commented on CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveUnusedAttributesPluginNGTest.java:
```diff
@@ -63,8 +65,10 @@ public void setUpMethod() {
-        PluginInteraction interaction = null;
-        PluginParameters parameters = null;
+
+        final PluginInteraction interaction = mock(PluginInteraction.class);
+        final PluginParameters parameters = mock(PluginParameters.class);
```
for `interaction`, you may want to include some handling of the functions that will get called on the mock when the plugin is run else you may get some unintended behaviour.
for `parameters` I would be tempted to just leave it as null as given that parameter isn't use, there is not much value in setting it to anything

This might be similar for other tests (and possibly other test classes)

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/PluginException.java:
```diff
@@ -70,6 +70,10 @@ public PluginException(PluginNotificationLevel notificationLevel, String message
         this.notificationLevel = notificationLevel;
     }
 
+    public PluginException() {	
+        throw new UnsupportedOperationException("Not supported yet.");
```
Was there a plan to implement this constructor? If not, this can probably be removed since there are other constructors (I couldn't see any usage on this one)

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginExecutionStageConstants.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class PluginExecutionStageConstants {
+    
+    private  PluginExecutionStageConstants() {
```
remove the extra space here

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -347,4 +366,90 @@ public PluginReport addChildReport(Plugin plugin) {
+     * @return a string representing a list of all the messages in the log.
+     */
+    public String logToString(Collection<String> log){
+        StringBuilder bld = new StringBuilder();
```
final

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReport.java:
```diff
@@ -347,4 +366,90 @@ public PluginReport addChildReport(Plugin plugin) {
+        
+        for (int i = 0; i < log.size(); ++i) {
+            if (i != 0){
+                bld.append("\n");
```
use `SeparatorConstants.NEWLINE`

antares1470 commented on CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/reporting/PluginReportUtilities.java:
```diff
@@ -0,0 +1,81 @@
+     * @param pluralForm
+     */
+    public static final String toString(final int count, final String singularForm, final String pluralForm) {
+        if (count == 1) {
+            return count + " " + singularForm;
+        } else {
+            return count + " " + pluralForm;
+        }
```
this can probably be a ternary statement (`return count + " " + count == 1 ? singularForm : pluralForm`)

capricornunicorn123 commented:
> The extra steps are probably most useful for developers if an issue is identified but its worth noting that if an error occurs, the plugin report will already reflect that in its output. Happy to hear other thoughts though.

The latest implementation tries to fit the needs of both types of users (those with an interest in the stage of plugin execution during runtime and those with an interest in what a plugin has ultimately completed). For long running scripts plugin reports will display detailed progress messages for the stage of plugin execution. Once complete the plugin report will only show relevant information regarding what the plug in did (i.e. modified, added or removed elements).

Can extend or refine this implementation further if it isn't appropriate.

antares1470 commented on CoreDataAccessView/test/unit/src/au/gov/asd/tac/constellation/views/dataaccess/plugins/clean/RemoveNodesPluginNGTest.java:
```diff
@@ -87,33 +88,32 @@ public void testCreateParameters() {
         
-        //Method should throw exception as Identifier attribute is not enabled on vertex
-        PluginExecution.withPlugin(instance).withParameters(parameters).executeNow(graph);
+        vertexIdentifierAttribute = VisualConcept.VertexAttribute.IDENTIFIER.ensure(graph);
```
is this one meant to be `SELECTED` given its meant to be testing the absence of the `IDENTIFIER` attribute?

antares1470 commented on CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/gui/NotifyDisplayerNGTest.java:
```diff
@@ -72,20 +72,26 @@ public static void tearDownClass() throws Exception {
-//    }
+    @Test
+    public void display() {    
+        if (!Boolean.TRUE.toString().equalsIgnoreCase(System.getProperty("java.awt.headless"))){
```
are you able to add a todo comment to address this for headless? I think this is fine as a temporary solution but I think it might be better long term to try get the test workable in the headless environment as well

antares1470 commented on CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/gui/NotifyDisplayerNGTest.java:
```diff
@@ -72,20 +72,26 @@ public static void tearDownClass() throws Exception {
-        displayWithIcon(true, false, true);
-        displayWithIcon(false, true, true);
-        displayWithIcon(false, false, false);
+        if (!Boolean.TRUE.toString().equalsIgnoreCase(System.getProperty("java.awt.headless"))){
```
same here
